### PR TITLE
Inkling Algorithm documentation

### PR DESCRIPTION
**Pull request by codemonkey9000**
* Added tables for all algorithm parameters
* Added example code for Inkling and syntax of clause within a curriculum vs. globally

PR Review Checklist:
- [x] Test search functionality on desktop
- [x] Test search functionality on mobile
- [x] Test filter functionality
- [x] Test mobile header navigation
- [x] Test desktop header navigation
- [x] Test desktop homepage navigation
- [ ] Run `npm run check-links` with 0 broken links
- [ ] Run `npm run write-good <files being edited>` to "lint" the docs

### CODE REVIEW

File: source/includes/inkling-reference/_algorithms.html.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+       lesson balancing
+           configure
+               constrain episode_length with Int8{-1},
+               constrain deque_size with UInt8{1}
+           until
+               maximize open_ai_gym_default_objective
+end
+```
+
+The algorithm clause is for advanced users who want to experiment changing machine learning algorithms and their training parameters. The clause provides syntax either within the curriculum statement to force a choice of algorithm for that curriculum, or it can be specified outside of a curriculum if used in multiple concepts (and therefore multiple lessons). It is also possible to override a global parameter inside of a curriculum if needed.
```
Comment (shnayder) [inline source/includes/inkling-reference/_algorithms.html.md]: I'd say
> ...and their _tuning_ parameters... 

The clause is always specified within a curriculum. If the user wants, they can define named algorithm settings they can re-use in multiple curricula.

File: source/includes/inkling-reference/_algorithms.html.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+* `TRPO` - estimation only
+
+
+### Shared Parameter Details
+
+The following parameters all apply to DQN, DDPG, Discrete DDPG, and TRPO. Additional parameter options are shown in subsequent tables below.
+
+| Parameter (Example use)      | Description |
+| -                            | -           |
+| `hidden_layer_size_descriptor` => [int, int, int, …] <br> (`hidden_layer_size_descriptor => [400, 300]` | An array of hidden layer: int is the hidden layer size. |
```
Comment (shnayder) [inline source/includes/inkling-reference/_algorithms.html.md]: "The sizes of each hidden layer. One integer size per hidden layer."

File: source/includes/inkling-reference/_algorithms.html.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+| `hidden_layer_activation_descriptor` => [string, string, string, …] <br> (`hidden_layer_activation_descriptor => [“relu”, “tanh”]`) | An array of activation functions (corresponding to the hidden layer size descriptor): string is the activation function type. Strings supported: “linear”, “tanh”, “relu”, “logistic”, “softmax”, “elu”. (Using this should be combined with `hidden_layer_size_descriptor`, in the given example “relu” is the activation function for the first layer, and “tanh” for the second one.) |
+| `conv_layer_descriptor` => “int1xint2:int3:int4:int5; …” <br> (`conv_layer_descriptor => “8x8:4:4:32; 3x3:1:1:16”`) | A semicolon delimited string of convolutional network configuration: int1 is x size, int2 is y size, int3 is x stride, int4 is y stride, int5 is number of filters. |
```
Comment (shnayder) [inline source/includes/inkling-reference/_algorithms.html.md]: Suggest replacing int1, etc with x_size, y_size, etc directly in the format.

File: source/includes/inkling-reference/_algorithms.html.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+| `samples_per_train` => int <br> (`samples_per_train => 10`) | The ratio of number of data sampling and training. |
+
+### Q-Table Parameters
+
+These are the only three parameters available for Q-Tables. They do not share any of the parameters from the first (shared) table.
+
+| Parameter (Example use)      | Description |
+| -                            | -           |
+| `exploration_decay` => float <br> (`exploration_decay => 0.00001`) | The exploration decay in DQN with the data type float. |
+| `q_learning_rate` => float <br> (`q_learning_rate => 0.0001`) | The learning rate for training the Q network. |
```
Comment (RuofanKong) [inline source/includes/inkling-reference/_algorithms.html.md]: I realized that this should be "The learning rate for training the Q Table." (not "Q network"), and I just fixed it in the google doc. Sorry for my typo...

File: source/includes/inkling-reference/_algorithms.html.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+| `hidden_layer_size_descriptor` => [int, int, int, …] <br> (`hidden_layer_size_descriptor => [400, 300]` | An array of hidden layer: int is the hidden layer size. |
+| `hidden_layer_activation_descriptor` => [string, string, string, …] <br> (`hidden_layer_activation_descriptor => [“relu”, “tanh”]`) | An array of activation functions (corresponding to the hidden layer size descriptor): string is the activation function type. Strings supported: “linear”, “tanh”, “relu”, “logistic”, “softmax”, “elu”. (Using this should be combined with `hidden_layer_size_descriptor`, in the given example “relu” is the activation function for the first layer, and “tanh” for the second one.) |
+| `conv_layer_descriptor` => “int1xint2:int3:int4:int5; …” <br> (`conv_layer_descriptor => “8x8:4:4:32; 3x3:1:1:16”`) | A semicolon delimited string of convolutional network configuration: int1 is x size, int2 is y size, int3 is x stride, int4 is y stride, int5 is number of filters. |
+| `conv_compression_size_descriptor` => [int, int, int, ...] <br> (`hidden_layer_descriptor => [20, 10]`)| An array of convolutional compressed size: int is the hidden layer size. |
```
Comment (shnayder) [inline source/includes/inkling-reference/_algorithms.html.md]: copy paste leftovers: `hidden_layer_descriptor` instead of `conv_compression_size_descriptor`. Should add a note about what this is. e.g. `If specified, will add one or more fully connected layers after the convolutional layers to compress the representation before incorporating other features`.

File: source/includes/inkling-reference/_algorithms.html.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+| `q_learning_rate` => float <br> (`q_learning_rate => 0.0001`) | The learning rate for training the Q network. |
+| `gamma` => float <br> (`gamma => 0.99`) | The discount rate for the Bellman Equation. |
+
+### TRPO-Specific Parameters
+
+In addition to the first (shared) table of parameters, TRPO also has one additional parameter available.
+
+| Parameter (Example use)      | Description |
+| -                            | -           |
+| `samples_per_train` => int <br> (`samples_per_train => 10`) | The ratio of number of data sampling and training. |
```
Comment (shnayder) [inline source/includes/inkling-reference/_algorithms.html.md]: would be nice to give default values and/or reasonable ranges. e.g. for this line, my impression for TRPO is that it normally uses ~10K samples per training, though perhaps I'm wrong.

File: source/includes/inkling-reference/_algorithms.html.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+### Q-Table Parameters
+
+These are the only three parameters available for Q-Tables. They do not share any of the parameters from the first (shared) table.
+
+| Parameter (Example use)      | Description |
+| -                            | -           |
+| `exploration_decay` => float <br> (`exploration_decay => 0.00001`) | The exploration decay in DQN with the data type float. |
```
Comment (shnayder) [inline source/includes/inkling-reference/_algorithms.html.md]: "DQN". 

Also, here and a few other places: the type is specified on the left, so no need to repeat in the description.

File: source/includes/inkling-reference/_algorithms.html.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+           until
+               maximize open_ai_gym_default_objective
+end
+```
+
+The algorithm clause is for advanced users who want to experiment changing machine learning algorithms and their training parameters. The clause provides syntax either within the curriculum statement to force a choice of algorithm for that curriculum, or it can be specified outside of a curriculum if used in multiple concepts (and therefore multiple lessons). It is also possible to override a global parameter inside of a curriculum if needed.
+
+### Valid Algorithm Identifiers
+
+* `DQN` (Deep Q Learning) - classification only
```
Comment (RuofanKong) [inline source/includes/inkling-reference/_algorithms.html.md]: Curious if we need full name for all algorithms. In case need it, I listed their names here :)

DDPG (Deep Deterministic Policy Gradients)
DiscreteDDPG (Deep Deterministic Policy Gradients with Discrete Actions)
Q-Table (Q Learning with Q Table)
TRPO (Trust Region Policy Optimization)

File: source/includes/inkling-reference/_algorithms.html.md (hunk 6 above)
Comment (codemonkey9000) [reply source/includes/inkling-reference/_algorithms.html.md]: @RuofanKong do you know what the default or range is for this?

Comment (RuofanKong) [reply source/includes/inkling-reference/_algorithms.html.md]: @codemonkey9000 By default, it uses 10K samples per training for TRPO in our code.

File: source/includes/inkling-reference/_algorithms.html.md (hunk 8 above)
Comment (codemonkey9000) [reply source/includes/inkling-reference/_algorithms.html.md]: It's very helpful, thank you Ruofan! I added them.

File: source/includes/inkling-reference/_algorithms.html.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+       lesson balancing
+           configure
+               constrain episode_length with Int8{-1},
+               constrain deque_size with UInt8{1}
+           until
+               maximize open_ai_gym_default_objective
+end
+```
+
+The algorithm clause is for advanced users who want to experiment changing machine learning algorithms and their tuning parameters. The clause is always specified within a curriculum. If the user wants, they can define named algorithm settings they can re-use in multiple curriculums (when using multiple concepts). It is also possible to override a global parameter inside of a curriculum if needed.
```
Comment (lucidbee) [inline source/includes/inkling-reference/_algorithms.html.md]: "experiment changing" should be "experiment with changing"

Comment (lucidbee) [reply source/includes/inkling-reference/_algorithms.html.md]: There is a global clause which is never specified within a curriculum and a curriculum algorithm clause so to say it is always specified in a curriculum is not quite right. Something like:
"Inkling supports an algorithm definition clause in global scope for use in multiple curriculums
as well as an algorithm use clause inside the curriculum statement which applies only to that curriculum."

File: source/includes/inkling-reference/_algorithms.html.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+simulator the_simulator(CartPoleConfig)
+   action (Action)
+   state (GameState)
+end
+
+
+algorithm My_TRPO_Settings
+   is TRPO
+   learning_rate => 5,
```
[the remark this answers is not in the document]
Comment (codemonkey9000) [reply source/includes/inkling-reference/_algorithms.html.md]: Tod confirmed that this does train. Going to publish for now and will follow up with Victor.
